### PR TITLE
feat: delayed invalid record case

### PR DESCRIPTION
**Pull request by sljoyy**
- When recording or replaying, the data needs to be cleaned up or marked invalid due to incomplete recording data.
- Since saving data is asynchronous, a delay is added to ensure that the cleanup is successful.

### CODE REVIEW

**Review by QizhengMo — APPROVED** [review]

File: arex-storage-web-api/src/main/java/com/arextest/storage/metric/AgentWorkingMetricService.java [hunk 1]
```diff
@@ -39,14 +39,18 @@ public class AgentWorkingMetricService {
...
   private final MockSourceEditionService editableService;
   private final InvalidReplayCaseService invalidReplayCaseService;
+  private final ScheduledExecutorService invalidCaseDelayedPool;
 
   public AgentWorkingMetricService(AgentWorkingService agentWorkingService,
                                    MockSourceEditionService editableService,
-      List<MetricListener> metricListeners, InvalidReplayCaseService invalidReplayCaseService) {
+                                   List<MetricListener> metricListeners,
+                                   InvalidReplayCaseService invalidReplayCaseService,
+                                   ScheduledExecutorService invalidCaseDelayedPool) {
```
Comment (mr3) [inline arex-storage-web-api/src/main/java/com/arextest/storage/metric/AgentWorkingMetricService.java:48]: `invalidReplayCaseService/invalidCaseDelayedPool` should not be in `MetricClass`

File: arex-storage-web-api/src/main/java/com/arextest/storage/metric/AgentWorkingMetricService.java [hunk 2]
```diff
@@ -132,5 +119,37 @@ private void invalidReplayCase(InvalidCaseRequest request) {
       metricListener.recordMatchingCount(METRIC_NAME, tags);
     }
   }
+  private class InvalidTask implements Runnable {
```
Comment (mr3) [inline arex-storage-web-api/src/main/java/com/arextest/storage/metric/AgentWorkingMetricService.java:122]: business prcessing code should not be in `MetricClass`,  the purpose of this class is just metric

File: arex-storage-web-api/src/main/java/com/arextest/storage/beans/ExecutorsConfiguration.java [hunk 3]
```diff
@@ -25,6 +25,14 @@ public ScheduledExecutorService coverageHandleDelayedPool() {
         new DiscardPolicy());
   }
 
+  @Bean
+  public ScheduledExecutorService invalidCaseDelayedPool() {
+    return new ScheduledThreadPoolExecutor(4,
+            createThreadFac("invalid-case-%d"),
+            new DiscardPolicy());
+  }
```
Comment (mr3) [inline arex-storage-web-api/src/main/java/com/arextest/storage/beans/ExecutorsConfiguration.java:33]: Recommended use `coverageHandleDelayedPool` directly

File: arex-storage-web-api/src/main/java/com/arextest/storage/web/controller/AgentRecordingController.java [hunk 4]
```diff
@@ -144,14 +144,24 @@ public Response saveTest(
   @PostMapping(value = "/invalidCase", produces = {MediaType.APPLICATION_JSON_VALUE})
   @ResponseBody
   public Response invalidCase(@RequestBody InvalidCaseRequest requestType) {
-      if (StringUtils.isEmpty(requestType.getRecordId())) {
-          LOGGER.warn("[[title=invalidCase]]agent invalid case recordId empty, {}", requestType);
-          return ResponseUtils.emptyRecordIdResponse();
-      }
+    try {
       MDCTracer.addRecordId(requestType.getRecordId());
       LOGGER.info("[[title=invalidCase]]agent invalid case, request:{}", requestType);
```
Comment (mr3) [inline arex-storage-web-api/src/main/java/com/arextest/storage/web/controller/AgentRecordingController.java:149]:
```java
if (StringUtils.isEmpty(requestType.getRecordId())) {
  LOGGER.warn("[[title=invalidCase]]agent invalid case recordId empty, {}", requestType);
  return ResponseUtils.emptyRecordIdResponse();
}

MDCTracer.addRecordId(requestType.getRecordId());
MDCTracer.addReplayId(requestType.getReplayId());
LOGGER.info("[[title=invalidCase]]agent invalid case, request:{}", requestType);


public static void addReplayId(String replayId) {
  if (StringUtils.isEmpty(replayId)) { return; }

  addAppType();
  put(REPLAY_ID, replayId);
}
```

File: arex-storage-web-api/src/main/java/com/arextest/storage/web/controller/AgentRecordingController.java [hunk 5]
```diff
@@ -144,14 +144,24 @@ public Response saveTest(
...
+        return ResponseUtils.emptyRecordIdResponse();
+      }
+      if (StringUtils.isNotEmpty(requestType.getReplayId())) {
+        MDCTracer.addReplayId(requestType.getRecordId());
+      }
       CompletableFuture.runAsync(() -> agentWorkingMetricService.invalidCase(requestType));
       return ResponseUtils.successResponse(true);
+    } catch (Throwable throwable) {
+        LOGGER.error("[[title=invalidCase]] invalidCase error:{}", throwable.getMessage(), throwable);
+      return ResponseUtils.exceptionResponse(throwable.getMessage());
```
Comment (mr3) [inline arex-storage-web-api/src/main/java/com/arextest/storage/web/controller/AgentRecordingController.java:161]: LOGGER.error("[[title=invalidCase]] invalidCase error:{}", throwable.getMessage());

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/InvalidCaseService.java [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+
+        @Override
+        public void run() {
+            MDCTracer.addRecordId(recordId);
+            try {
+                // recordId is not empty, means this is a record case
+                editableService.removeAll(ProviderNames.DEFAULT, recordId);
+            } catch (Exception exception) {
+                LOGGER.error("[[title=invalidCase]]invalid case error:{}", exception.getMessage(),
+                        exception);
```
Comment (mr3) [inline arex-storage-web-api/src/main/java/com/arextest/storage/service/InvalidCaseService.java:84]: LOGGER.error("[[title=invalidCase]]invalid case error:{}", exception.getMessage());

File: arex-storage-web-api/src/main/java/com/arextest/storage/service/AgentWorkingService.java [hunk 7]
```diff
@@ -175,4 +179,12 @@ public byte[] queryConfigFile(AREXMocker requestType) {
...
     return ZstdJacksonSerializer.EMPTY_INSTANCE;
   }
+
+  public void invalidCase(InvalidCaseRequest request) {
+    if (StringUtils.isNotEmpty(request.getReplayId())) {
+      invalidCaseService.invalidReplayCase(request.getReplayId());
+    } else {
+      invalidCaseService.invalidRecordCase(request.getRecordId());
+    }
+  }
```
Comment (mr3) [inline arex-storage-web-api/src/main/java/com/arextest/storage/service/AgentWorkingService.java:189]: put this method into InvalidCaseService

`invalidCaseService.invalidCase()`

File: arex-storage-web-api/src/main/java/com/arextest/storage/metric/AgentWorkingMetricService.java [hunk 8]
```diff
@@ -112,20 +88,16 @@ private void recordEntryTime(String path, AREXMocker item, long timeMillis) {
...
-    MDCTracer.addReplayId(replayId);
-    invalidReplayCaseService.saveInvalidCase(replayId);
-    LOGGER.info("[[title=invalidReplayCase]]invalid replayId:{}", replayId);
-    MDCTracer.clear();
-    // metric
     if (CollectionUtils.isEmpty(metricListeners)) {
       return;
     }
+    String path = StringUtils.isNotEmpty(request.getReplayId()) ? INVALID_REPLAY_CASE_METHOD_NAME
+        : INVALID_CASE_METHOD_NAME;
```
Comment (mr3) [inline arex-storage-web-api/src/main/java/com/arextest/storage/metric/AgentWorkingMetricService.java:97]: String scene = StringUtils.isNotEmpty(request.getReplayId()) ? "replay" : "record":

tags.put("scene", scene);

File: arex-storage-web-api/src/main/java/com/arextest/storage/metric/AgentWorkingMetricService.java [hunk 9]
```diff
@@ -112,20 +88,16 @@ private void recordEntryTime(String path, AREXMocker item, long timeMillis) {
     }
   }
 
-  private void invalidReplayCase(InvalidCaseRequest request) {
+  public void metricInvalidCase(InvalidCaseRequest request) {
```
Comment (mr3) [inline arex-storage-web-api/src/main/java/com/arextest/storage/metric/AgentWorkingMetricService.java:91]: recordInvalidCaseCount

**Review by mr3 — DISMISSED** [review]

**Review by QizhengMo — APPROVED** [review]

**Pull request merged**